src/Control/Effect.hs

```diff
@@ -25,20 +17,34 @@ module Control.Effect
 ) where
 
 import Control.Applicative (Alternative(..))
-import Control.Carrier
-import Control.Monad (ap, liftM)
+import Control.Monad (liftM, ap)
 import Control.Monad.Fail
 import Control.Monad.IO.Class
 import Prelude hiding (fail)
 
-data Eff effects a
-  = Return a
-  | Eff (effects (Eff effects) (Eff effects a))
+newtype Eff h a = Eff { unEff :: forall x . (a -> h x) -> h x }
```

Comment:
We no longer need a free-like `Eff` monad at all; instead, `Eff` is redefined as the codensity monad.